source/common/common/cancel_wrapper.h

```diff
@@ -0,0 +1,41 @@
+#pragma once
+
+#include <memory>
+#include <utility>
+
+#include "absl/functional/any_invocable.h"
+
+namespace Envoy {
+namespace CancelWrapper {
+
+using CancelFunction = absl::AnyInvocable<void()>;
+
+// Wraps a callback with a cancellation function. The cancellation function is saved
+// in an 'out' pointer argument to facilitate simple chaining in place, e.g. typical
+// usage would be of the form
+//
+// someFunctionThatTakesACallback(cancelWrapped([this](SomeResponse response) {
+//   doStuff(response);
+// }, &cancel_callback_));
+//
+// Then cancel_callback_ could be called (conditionally on having been populated) in the
+// class destructor to prevent use of "this" in a dispatched event that ends up occurring
```

Comment:
If it's intended to be called in the destructor, would it make more sense for the cancel function to be an object that by default cancels the operation in it's destructor? And maybe have a `clear()` or something on it to indicate that cancellation shouldn't happen?

If this works for the use case, you could also put the cancellation boolean into this returned object and not need a shared_ptr<bool>. 

I'm not sure if this idea actually works; just thinking of possibilities.

Reply:
You need the shared_ptr because the callback still needs access to it after the potential-canceller-side has destroyed it (unless you use weak_ptr and shared_ptr). If you use weak_ptr and shared_ptr then you lose the ability to discard the canceller-side and intentionally leave the callback active. If you use shared_ptr *and* wrap one side in an auto-cancel destructor then you've just made the non-cancellation case require the verbosity instead of the cancellation case requiring it, and also made it non-explicit so it's harder to follow what the intent is (a function for "don't cancel" is much more confusing than a function for "cancel")